Construct default CallSettings

geigerj opened the pull request:
Add a method to construct the default CallSettings dictionary from
information parsed from a yaml file.

tbetbetbe commented on google/gax/api_callable.py:
```diff
@@ -155,6 +156,107 @@ def inner(request):
+def construct_settings(
+        configuration, bundling_override, retry_override, retry_names, timeout):
+    """Constructs a dictionary mapping method names to CallSettings.
+
```
It'd be nice so include examples configuration yaml snippets in this docstring,

Have the yaml in the comment like that serves as a way of documenting the expected structure of the configuration dictionary. 

One yaml snippet might be enough, but it'd probably be nice to include distinct examples for each of the other arguments that might override the configuration.

geigerj replied:
Added a yaml snippet.

The way it's set up currently, users do not override using a yaml; rather, they construct native Python object (e.g., BundleOptions, RetryOptions) and associate them with method names in a dictionary. The yaml is a static file that shouldn't need to be seen/edited by the user.

tbetbetbe commented on test/test_api_callable.py:
```diff
@@ -42,6 +42,52 @@
+        }],
+    'retry_params': [
+        {
+            'name': 'default',
```
Are the types in the retry_params correct ?
Just checking, but I thought the YAML parser would not add quotes unless they were present in the yaml.

geigerj replied:
You're right; fixed.

tbetbetbe commented on google/gax/api_callable.py:
```diff
@@ -155,6 +156,107 @@ def inner(request):
 
 
+def _construct_bundling(method_config, method_bundling_override):
+    """Helper for ``construct_settings()``."""
```
Although private, this method has sufficient complexity that it warrants documenting the Args and Return value.

tbetbetbe commented on google/gax/api_callable.py:
```diff
@@ -155,6 +156,107 @@ def inner(request):
+
+
+def _construct_retry(
+        method_config, method_retry_override, retry_codes_def, retry_params,
```
Although private, this method has sufficient complexity that it warrants documenting the Args and Return value.

geigerj replied:
done

tbetbetbe commented on google/gax/api_callable.py:
```diff
@@ -155,6 +156,107 @@ def inner(request):
+            if method_bundling_override:
+                bundler = bundling.Executor(method_bundling_override)
+            else:
+                bundler = None
```
I think this would be easier to read if instead of nested if else block, here you used elif

tbetbetbe commented on google/gax/api_callable.py:
```diff
@@ -155,6 +156,107 @@ def inner(request):
+        method_config, method_retry_override, retry_codes_def, retry_params,
+        retry_names):
+    """Helper for ``construct_settings()``."""
+    if method_retry_override == OPTION_INHERIT:
```
This func looks simpler as: 

if method_retry_override != OPTION_INHERIT:
   return method_retry_override

...

geigerj commented:
Please take another look -- I've made a few more changes.

- corrected a pervasive misspelling of "discriminator"
- made minor changes to the yaml format
- added a temporary map of status codes to exceptions in config.py. I think that when we implement retry, we will need to check for the status code of failed calls, rather than the subclass of AbortionError, so this will need adjustment.

Also, the Travis build failed, but the error (pylint wrong-import-order error) seems to be spurious.

tbetbetbe commented:
LGTM
